Add HMAC authentication to checkSingle endpoint

jonathonherbert opened the pull request:
_PR 2/3 to to fulfil the workflow described in https://github.com/guardian/typerighter/discussions/344._

## What does this change?

Adds HMAC authentication to the checker service so we can authenticate HMAC headers for the `checkSingle` endpoint.

## Implementation notes

To do this, we needed to use `pan-domain-auth-play_2-8` in both the manager and checker. Previously, the checker was using `pan-domain-auth-validation`, as it only needs to verify cookies, and not create them. `panda-hmac-play_2-8` depends on `pan-domain-auth-play_2-8`, and we need hmac in our checker, so this PR expands the capabilities of the checker service slightly, allowing it to mint pan-domain-auth cookies.

This has a few benefits:
- testing the checker service no longer requires us getting a cookie from another `.local` service
- we can DRY up the code for managing authentication between services, which means that despite adding another authentication method in this PR, the diff removes lines.

I've leaned into that with a refactor that moves our various controller configurations into a single class, `PandaAuthController`, which configures both sorts of authentication – both services require the HMAC library and config, as both reading and writing HMAC headers requires access to panda secrets, so there's no unnecessary configuration.

## How to test

- All the current endpoints should authenticate as expected. Running locally or in CODE, you should be able to access both the manager and checker service as before – but only when authenticated. Try removing your cookies for each service and hitting an authenticated endpoint (everything that isn't a healthcheck.) You should go through an oAuth flow and sign in as usual.
- The HMAC authentication should work as expected. Testing this is a little like #363, but ... hard mode! Following the steps in that PR, we'll also need to sign our request with HMAC headers.

We can do that by sending a request as per #363, but we'll need to sign the request with HMAC headers. I've added a script to do that in Postman below. The secret is in AWS Secrets Manager in the usual account – speak to me if it's not clear how to get hold of it. You'll need the secret for the `DEV` stage.

Alternatively, in the next PR in the series, the rule manager will use this authentication to talk to the checker service directly, so you could use that PR to verify _this_ PR's behaviour. 🙃 

<details>
<summary>Postman script</summary>

```javascript
const URL = require("url");

const uri = pm.request.url.toString();
const secret = "changeme";

const httpDate = new Date().toUTCString();
const path = new URL.parse(uri).path;
const stringToSign = `${httpDate}\n${path}`;
const stringToSignBytes = CryptoJS.enc.Utf8.parse(stringToSign);
const secretBytes = CryptoJS.enc.Utf8.parse(secret);

const signature = CryptoJS.enc.Base64.stringify(CryptoJS.HmacSHA256(stringToSignBytes, secretBytes));
const authToken = `HMAC ${signature}`;
console.log(`${authToken}, ${httpDate}`);

pm.request.headers.add({ key: 'X-Gu-Tools-HMAC-Date', value: httpDate });
pm.request.headers.add({ key: 'X-Gu-Tools-HMAC-Token', value: authToken });
```

</details>

Fweddi commented on build.sbt:
```diff
@@ -78,6 +84,7 @@ val commonLib = (project in file(s"$appsFolder/common-lib"))
   .settings(
     commonSettings,
     libraryDependencies ++= Seq(
+      ws,
```
This is defined for play projects in https://github.com/guardian/typerighter/blob/jsh/add-hmac/build.sbt#L100, do we need it in the commonLib?

jonathonherbert replied:
Thanks for spotting! fb3989918bb29fae1e506737fc94d408c66acdac

Fweddi commented on build.sbt:
```diff
@@ -57,16 +57,22 @@ val commonSettings = Seq(
     "com.google.apis" % "google-api-services-sheets" % "v4-rev20221216-2.0.0",
     "org.languagetool" % "languagetool-core" % languageToolVersion,
     "org.languagetool" % "language-en" % languageToolVersion,
+    "com.gu" %% "content-api-models-scala" % capiModelsVersion,
+    "com.gu" %% "content-api-models-json" % capiModelsVersion,
+    "com.gu" %% "content-api-client-aws" % "0.7",
+    "com.gu" %% "content-api-client-default" % capiClientVersion,
```
These all look compatible 👍

Fweddi commented on build.sbt:
```diff
@@ -57,16 +57,22 @@ val commonSettings = Seq(
     "org.mockito" %% "mockito-scala-scalatest" % "1.17.12",
     "com.gu" %% "simple-configuration-ssm" % "1.5.7",
-    "com.gu" %% "pan-domain-auth-verification" % "1.2.1",
+    "com.gu" %% "pan-domain-auth-play_2-8" % "1.2.1",
```
panda-hmac-play_2-8 uses 1.2.0 which is the same minor version 👍

Fweddi reviewed (changes requested):
When I delete my cookie and head to the checker service, it gives a 401. It doesn't seem to issue a Panda cookie as expected.

For the HMAC request, I sent a request to /checkSingle using the Postman instructions you supplied, and got the following error message:

```json 
{
    "error": {
        "id": "82adkec66",
        "requestId": 8,
        "exception": {
            "title": "Execution exception",
            "description": "[SdkClientException: Unable to load AWS credentials from any provider in the chain: [EnvironmentVariableCredentialsProvider: Unable to load AWS credentials from environment variables (AWS_ACCESS_KEY_ID (or AWS_ACCESS_KEY) and AWS_SECRET_KEY (or AWS_SECRET_ACCESS_KEY)), SystemPropertiesCredentialsProvider: Unable to load AWS credentials from Java system properties (aws.accessKeyId and aws.secretKey), WebIdentityTokenCredentialsProvider: To use assume role profiles the aws-java-sdk-sts module must be on the class path., com.amazonaws.auth.profile.ProfileCredentialsProvider@74a742df: No AWS profile named 'default', com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper@393e7c1a: Failed to connect to service endpoint: ]]",
            "stacktrace": [
                "com.amazonaws.SdkClientException: Unable to load AWS credentials from any provider in the chain: [EnvironmentVariableCredentialsProvider: Unable to load AWS credentials from environment variables (AWS_ACCESS_KEY_ID (or AWS_ACCESS_KEY) and AWS_SECRET_KEY (or AWS_SECRET_ACCESS_KEY)), SystemPropertiesCredentialsProvider: Unable to load AWS credentials from Java system properties (aws.accessKeyId and aws.secretKey), WebIdentityTokenCredentialsProvider: To use assume role profiles the aws-java-sdk-sts module must be on the class path., com.amazonaws.auth.profile.ProfileCredentialsProvider@74a742df: No AWS profile named 'default', com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper@393e7c1a: Failed to connect to service endpoint: ]",
                "at com.amazonaws.auth.AWSCredentialsProviderChain.getCredentials(AWSCredentialsProviderChain.java:142)",
                "at com.amazonaws.http.AmazonHttpClient$RequestExecutor.getCredentialsFromContext(AmazonHttpClient.java:1269)",
                "at com.amazonaws.http.AmazonHttpClient$RequestExecutor.runBeforeRequestHandlers(AmazonHttpClient.java:845)",
                "at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:794)",
                "at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:781)",
                "at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:755)",
                "at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:715)",
                "at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:697)",
                "at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:561)",
                "at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:541)",
                "at com.amazonaws.services.secretsmanager.AWSSecretsManagerClient.doInvoke(AWSSecretsManagerClient.java:2770)",
                "at com.amazonaws.services.secretsmanager.AWSSecretsManagerClient.invoke(AWSSecretsManagerClient.java:2737)",
                "at com.amazonaws.services.secretsmanager.AWSSecretsManagerClient.invoke(AWSSecretsManagerClient.java:2726)",
                "at com.amazonaws.services.secretsmanager.AWSSecretsManagerClient.executeGetSecretValue(AWSSecretsManagerClient.java:1089)",
                "at com.amazonaws.services.secretsmanager.AWSSecretsManagerClient.getSecretValue(AWSSecretsManagerClient.java:1058)",
                "at com.gu.typerighter.lib.CommonConfig.$anonfun$hmacSecrets$1(CommonConfig.scala:70)",
                "at scala.collection.immutable.List.map(List.scala:246)",
                "at com.gu.typerighter.lib.CommonConfig.hmacSecrets$lzycompute(CommonConfig.scala:64)",
                "at com.gu.typerighter.lib.CommonConfig.hmacSecrets(CommonConfig.scala:64)",
                "at com.gu.typerighter.controllers.PandaAuthController.secretKeys(PandaAuthController.scala:42)",
                "at com.gu.pandahmac.HMACSecrets.gatherSecrets(HmacSecrets.scala:13)",
                "at com.gu.pandahmac.HMACSecrets.gatherSecrets$(HmacSecrets.scala:12)",
                "at com.gu.typerighter.controllers.PandaAuthController.gatherSecrets(PandaAuthController.scala:12)",
                "at com.gu.pandahmac.HMACSecrets.validateHMACHeaders(HmacSecrets.scala:20)",
                "at com.gu.pandahmac.HMACSecrets.validateHMACHeaders$(HmacSecrets.scala:19)",
                "at com.gu.typerighter.controllers.PandaAuthController.validateHMACHeaders(PandaAuthController.scala:12)",
                "at com.gu.pandahmac.HMACAuthActions.com$gu$pandahmac$HMACAuthActions$$authByKeyOrPanda(HmacAuthActions.scala:39)",
                "at com.gu.pandahmac.HMACAuthActions$APIHMACAuthAction$.invokeBlock(HmacAuthActions.scala:79)",
                "at com.gu.pandahmac.HMACAuthActions$APIHMACAuthAction$.invokeBlock(HmacAuthActions.scala:74)",
                "at play.api.mvc.ActionBuilder$$anon$8.invokeBlock(Action.scala:289)",
                "at play.api.mvc.ActionBuilder$$anon$8.invokeBlock(Action.scala:284)",
                "at play.api.mvc.ActionBuilder$$anon$9.apply(Action.scala:379)",
                "at play.api.mvc.Action.$anonfun$apply$4(Action.scala:82)",
                "at play.api.libs.streams.StrictAccumulator.$anonfun$mapFuture$4(Accumulator.scala:168)",
                "at scala.util.Try$.apply(Try.scala:210)",
                "at play.api.libs.streams.StrictAccumulator.$anonfun$mapFuture$3(Accumulator.scala:168)",
                "at scala.Function1.$anonfun$andThen$1(Function1.scala:85)",
                "at scala.Function1.$anonfun$andThen$1(Function1.scala:85)",
                "at scala.Function1.$anonfun$andThen$1(Function1.scala:85)",
                "at play.api.libs.streams.StrictAccumulator.run(Accumulator.scala:200)",
                "at play.core.server.AkkaHttpServer.$anonfun$runAction$4(AkkaHttpServer.scala:429)",
                "at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)",
                "at akka.http.scaladsl.util.FastFuture$.$anonfun$transformWith$3(FastFuture.scala:51)",
                "at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:484)",
                "at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)",
                "at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)",
                "at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)",
                "at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)",
                "at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)",
                "at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)",
                "at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)",
                "at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)",
                "at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)",
                "at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)",
                "at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)"
            ]
        }
    }
}
```

It looks to me like it's looking for the default profile, rather than `composer`.

jonathonherbert commented:
Thanks for the feedback, @Fweddi! 

_> When I delete my cookie and head to the checker service, it gives a 401. It doesn't seem to issue a Panda cookie as expected._

Oof, I'd forgotten to do the work to wire up the oAuth service for the checker project. That's done and tested locally in [bd82176](https://github.com/guardian/typerighter/pull/383/commits/bd821764c568093d0a12e644cb507b3df9a5dd72). I've added the projects to Google Cloud Services a few minutes ago – it may take a little time for that to propagate.

_> I sent a request to /checkSingle using the Postman instructions you supplied, and got the following error message:_

That should be fixed in [1ada6ac](https://github.com/guardian/typerighter/pull/383/commits/1ada6ac688c5d1392976a6590ab1fa09e4a2aa6e) – it's a change I'd made in #389, but failed to backport as I split this work into separate PRs. Now tested locally – let me know if you have more success!
